### DISCUSSION TITLE
Do moon repo offer lifecycle scripts?

### DISCUSSION THREAD

**Discussion by diraneyya** [moonrepo/moon#806] [Q&A] [answered]
Hello there 👋

I am a full-stack web development teacher and I got to hear about this amazing project from the TLDR newsletter. I would like to start by thanking everyone involved for making such a wonderful tool.

I am trying to help people who just started programming scaffold a full-stack (i.e. a frontend/backend) monorepo using ExpressJS for the backend and ReactJS for the frontend.

We are using Prisma as the ORM for abstracting access to relational-model databases.

One of my goals is to be able to share the Prisma-client types with the frontend. Both of the frontend and the backend use Typescript.

We are using Rome for both the frontend and the backend with common (i.e. workspace-level) linting and formatting tasks.

Aside from stating the goal relevant to my question, my question is about automating certain tasks to run when the workspace is setup (either using `moon setup`) or when running any of the tasks. As this is where I am stuck at the moment on my way towards my goal.

This feature is called _lifecycle scripts_ in npm and includes _pre and post scripts_ as well as others, such as the _postinstall_ script which I am after (for _generating the Prisma client_ and _deploying the database_).

I would be keen to learn more about such features in moon if they exist.

I really absolutely love this tool 💖

> **PS:** I just now took a look at [this](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#generating-prisma-client-in-the-postinstall-hook-of-prismaclient) and hence I am not sure if my question is still relevant. Namely, it might no longer be relevant to moon to offer a "postinstall" script but I am keeping the question since the _pre and post scripts_ are still relevant to a task runner.

**Comment by milesj** [comment, marked as answer]
@diraneyya Great question! Thanks for the in-depth discussion. At this time we do not support life-cycles, but it is definitely on our roadmap. Locking down what the API would look like is the hardest part.

However, there is an approach that you could possible use. If you want a global "setup everything" kind of task, you can use `implicitDeps`: https://moonrepo.dev/docs/config/tasks#implicitdeps 

With this, you can define a task in `.moon/tasks.yml`, and have it be inherited by _every_ task as a dependency. So anytime a task is ran, this setup task will be also. Something like this:

```yaml
implicitDeps:
  - '~:globalSetup'

tasks:
  globalSetup:
    command: 'do something'
```

This _should_ work how you want, but we can definitely prioritize life-cycles.
